dns/resolver.go

```diff
@@ -0,0 +1,385 @@
+		},
+	}
+}
+
+// NewTLSResolver creates a [Resolver] that implements the [DNS-over-TLS] protocol, using a [transport.StreamDialer]
+// to connect to the resolverAddr, and the resolverName as the TLS server name.
+// It creates a new connection to the resolver for every request.
+//
+// [DNS-over-TLS]: https://datatracker.ietf.org/doc/html/rfc7858
+func NewTLSResolver(sd transport.StreamDialer, resolverAddr string, resolverName string) Resolver {
+	resolverAddr = ensurePort(resolverAddr, "853")
+	return &streamResolver{
+		NewConn: func(ctx context.Context) (transport.StreamConn, error) {
+			baseConn, err := sd.Dial(ctx, resolverAddr)
+			if err != nil {
+				return nil, err
+			}
+			return tls.WrapConn(ctx, baseConn, resolverName)
+		},
+	}
+}
+
+// NewHTTPSResolver creates a [Resolver] that implements the [DNS-over-HTTPS] protocol, using a [transport.StreamDialer]
+// to connect to the resolverAddr, and the url as the DoH template URI.
+// It uses an internal HTTP client that reuses connections when possible.
+//
+// [DNS-over-HTTPS]: https://datatracker.ietf.org/doc/html/rfc8484
+func NewHTTPSResolver(sd transport.StreamDialer, resolverAddr string, url string) Resolver {
+	resolverAddr = ensurePort(resolverAddr, "443")
+	dialContext := func(ctx context.Context, network, addr string) (net.Conn, error) {
+		if !strings.HasPrefix(network, "tcp") {
+			// TODO: Support UDP for QUIC.
+			return nil, fmt.Errorf("protocol not supported: %v", network)
+		}
+		conn, err := sd.Dial(ctx, resolverAddr)
+		if err != nil {
+			return nil, &nestedError{ErrDial, err}
+		}
+		return conn, nil
+	}
+	// TODO: add mechanism to close idle connections.
+	// Copied from Intra: https://github.com/Jigsaw-Code/Intra/blob/d3554846a1146ae695e28a8ed6dd07f0cd310c5a/Android/tun2socks/intra/doh/doh.go#L213-L219
+	httpClient := http.Client{
+		Transport: &http.Transport{
+			DialContext:           dialContext,
+			ForceAttemptHTTP2:     true,
+			TLSHandshakeTimeout:   10 * time.Second,
+			ResponseHeaderTimeout: 20 * time.Second, // Same value as Android DNS-over-TLS
+		},
+	}
+	return FuncResolver(func(ctx context.Context, q dnsmessage.Question) (*dnsmessage.Message, error) {
+		// Prepare request.
+		buf, err := appendRequest(0, q, make([]byte, 0, 512))
+		if err != nil {
+			return nil, &nestedError{ErrBadRequest, fmt.Errorf("append request failed: %w", err)}
+		}
+		httpReq, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewBuffer(buf))
+		if err != nil {
+			return nil, &nestedError{ErrBadRequest, fmt.Errorf("create HTTP request failed: %w", err)}
+		}
+		const mimetype = "application/dns-message"
+		httpReq.Header.Add("Accept", mimetype)
+		httpReq.Header.Add("Content-Type", mimetype)
+
+		// Send request and get response.
+		httpResp, err := httpClient.Do(httpReq)
+		if err != nil {
+			return nil, &nestedError{ErrReceive, fmt.Errorf("failed to get HTTP response: %w", err)}
+		}
+		defer httpResp.Body.Close()
+		if httpResp.StatusCode != http.StatusOK {
```

Comment:
Do we need to handle [`301` response](https://developers.google.com/speed/public-dns/docs/doh#redirections) here?

Reply:
The http.Client handles redirects by default: https://pkg.go.dev/net/http#Client